### ISSUE TITLE
Documentation - How to communicate between the background task (isolate) and other widgets

### ISSUE THREAD

**Issue by jonathanMNg** [fluttercommunity/flutter_workmanager#541]
- [x] I have read the [README](https://github.com/vrtdev/flutter_workmanager/blob/master/README.md)
- [x] I have done the setup for [Android](https://github.com/vrtdev/flutter_workmanager/blob/master/ANDROID_SETUP.md)
- [x] I have done the setup for [iOS](https://github.com/vrtdev/flutter_workmanager/blob/master/IOS_SETUP.md)
- [x] I have ran the sample app and it does not work there

**Version**
0.5.0

**Describe the error**
Hi,
I'm trying to execute a Bloc event with the workmanager callbackDispatcher().
However, because the function callbackDispatcher itself is an Isolate, it can't modify the variables of the main execution. My workaround is to use SendPort and ReceivePort.

In the main.dart:

```
@pragma('vm:entry-point')
void callbackDispatcher() {
  Workmanager().executeTask((task, inputData) {
    if(task == 'sleep-timer') {
      final SendPort? send = IsolateNameServer.lookupPortByName('sleep_timer_send_port');
      if (send != null) {
        send.send(inputData);
      }
    }
    return Future.value(true);
  });
}

Future<void> main() async {
  WidgetsFlutterBinding.ensureInitialized();
  await Workmanager().initialize(callbackDispatcher,isInDebugMode: true);
  runApp(const MyApp());
}

```
In the timer_bloc.dart:
```

 TimerBloc() : super(TimerInitial()) {

    on<Start>((event, emit) async {
      Workmanager().cancelByTag('sleep-timer-tag');
      final ReceivePort port = ReceivePort();
      _bindBackgroundIsolate(port);
      await Workmanager().registerOneOffTask(
          UniqueKey().toString(),
          'sleep-timer',
          tag: 'sleep-timer-tag',
          initialDelay: event.duration,
          inputData: {"executeTask": true},
          existingWorkPolicy: ExistingWorkPolicy.replace
      );
  });

  void _bindBackgroundIsolate(ReceivePort port) {
    final isSuccess = IsolateNameServer.registerPortWithName(port.sendPort, 'sleep_timer_send_port');
    if (!isSuccess) {
      _unbindBackgroundIsolate('sleep_timer_send_port');
      _bindBackgroundIsolate(port);
      return;
    }
    port.listen((dynamic data) async {
      // final isDone = data;
      if(data != null && data['executeTask']) {
        do_other_task();
      }
    });
  }
  void _unbindBackgroundIsolate(String name) {
    IsolateNameServer.removePortNameMapping(name);
  }
}
```

I think that adding this to the documentation would help others. FYI, I copied that code from this project [flutter_downloader](https://pub.dev/packages/flutter_downloader)
Thanks,

**Comment by SEGVeenstra** [comment]
I can imagine this being a very common use-case when using background tasks.
At least I found out I needed this within the first week of using background tasks.

Not sure if the documentation would require a full guide or maybe just a section with some references to this issue for example, as this is more about how Isolates work then about this package specifically.

Anyway, I'm very grateful that you shared this!

**Comment by aap01** [comment]
@jonathanMNg 
Thanks!
Very resourceful. Had I found this in the documentation, I would have saved 4 hours of my day.

**Comment by PsyOhm23** [comment]
Thank you but unfortunately this solution don't work after killing the app or restarting the device :(

**Comment by jonathanMNg** [comment]
> Thank you but unfortunately this solution don't work after killing the app or restarting the device :(

Yes, there is no way to do it in IOS, and you maybe able to do it in Android if you can fork the Android kernel and modify it to add your app to run after killing/restarting :)

**Comment by ened** [comment]
Yes — this is exactly the pattern to use, and it's a very common need. Here's the short version:

**Background isolate → UI isolate:** create a `ReceivePort` in `main()`, pass its `SendPort` into the background isolate (e.g. via `inputData` or a shared store), and have the callback `send()` messages back. The UI listens on the `ReceivePort` and updates widgets.

```dart
void main() {
  final receivePort = ReceivePort();
  receivePort.listen((message) {
    // Update UI state with the message
  });
  Workmanager().initialize(callbackDispatcher,
      inputData: <String, dynamic>{'sendPort': receivePort.sendPort});
  runApp(MyApp());
}
```

Note that `SendPort`/`ReceivePort` only work **while the main isolate is alive**. If the app was killed or the OS terminated it, the background task runs with no UI to talk to — the callback can still do its work (persist data, post a notification, schedule the next task), but it cannot update widgets. The commenter above is right: there is no supported way to reach the UI after the app has been force-quit, on either platform. Persist results (SharedPreferences, database, notifications) and let the UI pick them up on next launch.

Docs: the Quick Start now explains the separate-isolate model (see the `callbackDispatcher` section in #688). If you'd like a dedicated guide with full SendPort/ReceivePort examples, that's a good candidate for a docs PR — happy to review one.

**Comment by ened** [comment]
Answered by the maintainer above with the exact pattern: create a `ReceivePort` in `main()`, pass its `SendPort` into the background isolate, and have the callback `send()` messages back to the UI. There is also a great community summary in this thread with working code. Closing as answered (this is standard Flutter isolate communication, not a package API gap).
